### PR TITLE
Updated mailroom.py with OO principles and generated tests

### PR DESCRIPTION
**Pull request by shibinmathew**
mailroom.py and test_mailroom.py for lesson 9.

### CODE REVIEW

File: students/shibin_mathew/lesson9/mailroom.py [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+    def names(self, names):
+        self._names = names
+
+    @property
+    def amount(self):
+        return self._amount
+
+    @amount.setter
+    def amount(self, amount):
+        self._amount = amount
```
Comment (natasha-aleksandrova) [inline students/shibin_mathew/lesson9/mailroom.py]: if you are not doing anything special you don't really need to use properties, just regular class attributes should work.

File: students/shibin_mathew/lesson9/mailroom.py [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+        for names in donors2:
+            list_test.append(names)
+        return list_test
+
+    def send_thank_you(self, donors2):
+        #  Function to Send a Thank You Email
+        c1 = Collection()
+        program_quit = False
+        while not program_quit:
+            name = input("Please input your full name\n")
```
Comment (natasha-aleksandrova) [inline students/shibin_mathew/lesson9/mailroom.py]: remember the separation of concern from the assignment page, you should not have any `input` functions in your data handling code

File: students/shibin_mathew/lesson9/mailroom.py [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+        return self._amount
+
+    @amount.setter
+    def amount(self, amount):
+        self._amount = amount
+
+
+class Collection:
+    def __init__(self, donors2=None):
+        self.donors2 = donors2
```
Comment (natasha-aleksandrova) [inline students/shibin_mathew/lesson9/mailroom.py]: what does `donors2` mean?

File: students/shibin_mathew/lesson9/mailroom.py [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+    def create_report(self, donors2):
+        #  Function to create a report
+        list_test = []
+        print("hello")
+        print("{: <20s}{: ^4s}{: <10s}{: ^4s}{: <10s}{: ^4s}{: <10s}"
+              .format('Donor Name', '|', 'Total Given', '|',
+                      'Num Gifts', '|', 'Average Gift'))
+        print('-' * 70)
+        print(' ')
+        for names in donors2:
```
Comment (natasha-aleksandrova) [inline students/shibin_mathew/lesson9/mailroom.py]: the main point of the class is to use the attribute (`self.donors2`), why is this accepting it as a param instead?

File: students/shibin_mathew/lesson9/mailroom.py [hunk 5]
```diff
@@ -0,0 +1,152 @@
...
+    d3 = Donor("Jordy", 12)
+    d4 = Donor("Andreck", 14)
+    list_ = [(d1.names, d1.amount), (d2.names, d2.amount), (d3.names, d3.amount), (d4.names, d4.amount)]
+    donors2 = defaultdict(list)
+    for name, amount in list_:
+        donors2[name].append(amount)
+
+    q = Collection()
+    q.donors2 = donors2
+    q.question_module(q.donors2)
```
Comment (natasha-aleksandrova) [inline students/shibin_mathew/lesson9/mailroom.py]: couple points here:
* `Donor` class should have methods that deal with single donor like letter text for example
* Your collection class should deal with multiple donors, so it should have access to your donor objects something like `collection = Collection([Donor("name1", 100), Donor("name2", 200)])`, you should be doing all of the operations on that collection of donors (not creating more instances of collection class within the collection class?)
* actual data structure for donors should be hidden in your `Collection` class (lines 141 - 144)
* input functions should be outside of data handling classes (Donor and Collection)

File: students/shibin_mathew/lesson9/mailroom.py [hunk 6]
```diff
@@ -0,0 +1,152 @@
...
+
+
+class Donor:
+    def __init__(self, names, amount):
+        self.names = names
+        self.amount = amount
+
+    @property
+    def names(self):
+        return "{}".format(self._names)
```
Comment (natasha-aleksandrova) [inline students/shibin_mathew/lesson9/mailroom.py]: don't need this as a property, `_names` is already a string so no need to format.

also, it shouldn't be `names`, but a single `name`

File: students/shibin_mathew/lesson9/mailroom.py (hunk 1 above)
Comment (shibinmathew) [reply students/shibin_mathew/lesson9/mailroom.py]: Good point - I made the change.

File: students/shibin_mathew/lesson9/mailroom.py (hunk 2 above)
Comment (shibinmathew) [reply students/shibin_mathew/lesson9/mailroom.py]: I created a class to handle the question module which will separate the functions that take in the inputs with the functions that handle the data.

File: students/shibin_mathew/lesson9/mailroom.py (hunk 3 above)
Comment (shibinmathew) [reply students/shibin_mathew/lesson9/mailroom.py]: passing in a donors list - I would assume in the industry you can access this from a database that holds the donor information.

File: students/shibin_mathew/lesson9/mailroom.py (hunk 4 above)
Comment (shibinmathew) [reply students/shibin_mathew/lesson9/mailroom.py]: Good point - this was my mistake. I made the change.

**Comment by shibinmathew** [comment]
I refactored a lot of the code to implement your suggestions.

**Pull request merged**